universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts

```diff
@@ -0,0 +1,46 @@
+import {BalanceObserver} from './BalanceObserver';
+import {TokenDetailsWithBalance} from '@universal-login/commons';
+import {utils} from 'ethers';
+
+export interface PriceOracle {
+  getTokenPrice: (tokenSymbol: string) => Promise<number>;
+}
+
+export class AggregateBalanceObserver {
+  private callbacks: Function[] = [];
+  private unsubscribeBalanceObserver?: Function;
+  constructor(private balanceObserver: BalanceObserver, private priceOracle: PriceOracle) {}
+
+  subscribe(callback: Function) {
+    this.callbacks.push(callback);
+    if (this.callbacks.length === 1) {
+      this.unsubscribeBalanceObserver = this.balanceObserver.subscribe(this.callbackWrapper.bind(this));
+    }
+
+    const unsubscribe = () => {
+      this.callbacks = this.callbacks.filter((element) => callback !== element);
+      if (this.callbacks.length === 0) {
+        this.unsubscribeBalanceObserver!();
+      }
+    };
+    return unsubscribe;
+  }
+
+  async callbackWrapper(tokensDetailsWithBalance: TokenDetailsWithBalance[]) {
+    const aggregatedBalanceInUSD = await this.getAggregatedBalanceInUSD(tokensDetailsWithBalance);
+    this.callbacks.forEach((callback) => callback(aggregatedBalanceInUSD));
+  }
+
+  async getAggregatedBalanceInUSD(tokensDetailsWithBalance: TokenDetailsWithBalance[]) {
+    return tokensDetailsWithBalance.reduce(
+      async (total, tokenDetails) => (await total + await this.getTokenBalanceInUSD(tokenDetails)),
```

Comment:
`await total`, did you mean just `total`?

Reply:
Ok, got it. accum is a promise :) clever. 👍

Reply:
Or is it? I think it should be more like:
`Promise.resolve(await total + await this.getTokenBalanceInUSD(tokenDetails));`